hashes/src/hkdf.rs

```diff
@@ -0,0 +1,143 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! HMAC-based Extract-and-Expand Key Derivation Function (HKDF).
+//!
+//! Implementation based on RFC5869, but the interface is scoped
+//! to BIP324's requirements.
+
+use core::fmt;
+
+use crate::{Hash, HashEngine, Hmac, HmacEngine};
+
+// Output keying material max length multiple.
+const MAX_OUTPUT_BLOCKS: usize = 255;
+
+/// Size of output exceeds maximum length allowed.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct MaxLengthError {
+    max: usize,
+}
+
+impl fmt::Display for MaxLengthError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "exceeds {} byte max output material limit", self.max)
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for MaxLengthError {}
+
+/// HMAC-based Extract-and-Expand Key Derivation Function (HKDF).
+pub struct Hkdf<T: Hash> {
+    /// Pseudorandom key based on the extract step.
+    prk: Hmac<T>,
+}
+
+impl<T: Hash> Hkdf<T> {
+    /// Initialize a HKDF by performing the extract step.
+    pub fn new(salt: &[u8], ikm: &[u8]) -> Self {
+        let mut hmac_engine: HmacEngine<T> = HmacEngine::new(salt);
+        hmac_engine.input(ikm);
+        Self { prk: Hmac::from_engine(hmac_engine) }
+    }
+
+    /// Expand the key to generate output key material in okm.
+    pub fn expand(&self, info: &[u8], okm: &mut [u8]) -> Result<(), MaxLengthError> {
```

Comment:
What is the benefit in having this mutate `okm` instead of allocating and returning a new vector? Is it performance only? Is `expand` expected to be called repeatedly or only ever once?

Also, we probably want to do some [sanity] length checks on `okm` - this panics
```rust
    #[test]
    #[cfg(feature = "alloc")]
    fn panic() {
        let salt = Vec::from_hex("000102030405060708090a0b0c").unwrap();
        let ikm = Vec::from_hex("0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b").unwrap();
        let info = Vec::from_hex("f0f1f2f3f4f5f6f7f8f9").unwrap();

        let hkdf = Hkdf::<sha256::Hash>::new(&salt, &ikm);
        let mut okm = [0u8; 1];
        hkdf.expand(&info, &mut okm).unwrap();
    }
```

Reply:
We have been very conservative with our std and alloc usage just based on other implementations we have seen and to keep it as agnostic as possible. But we could have been to strict in this case?

Either way, I'll definitely add some more assertions and unit tests.

Reply:
I think "avoiding an `alloc` dependency" is a great reason. But we should probably also have an `alloc`-enabled API which maybe wraps this with something that returns a `Vec`.

Reply:
Gave a little wrapper a shot, let me know if that is what you had in mind.

Reply:
LGTM!